cpp/src/decisiontree/levelalgo/common_kernel.cuh

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#define LEAF 0xFFFFFFFF
+#define PUSHRIGHT 0x00000001
+
+__global__ void setup_counts_kernel(unsigned int* sample_cnt,
+                                    const unsigned int* __restrict__ rowids,
+                                    const int n_sampled_rows) {
+  int threadid = threadIdx.x + blockIdx.x * blockDim.x;
+  for (int tid = threadid; tid < n_sampled_rows;
+       tid += blockDim.x * gridDim.x) {
+    unsigned int stid = rowids[tid];
+    atomicAdd(&sample_cnt[stid], 1);
```

Comment:
Why use `atomicAdd`? Do we expect a particular row to be sampled more than once?

Reply:
yes That is usually the case when you bootstrap data from random forest,

Reply:
ah.. so this count is not for a single tree, but for the whole forest?

Reply:
No its still for single tree, But counts vary for each tree.

Reply:
Sorry, I'm confused now. How can bootstrap lead to duplicate input rows for a given tree?

Reply:
Check here. This is also the case for the old algo as well. As such in general definition of RF, involves bootstrap with replacement. Check here https://github.com/rapidsai/cuml/blob/a053eefb9b8d90f900ef32b53c7ad55da42f4316/cpp/src/randomforest/randomforest_impl.cuh#L71